### ISSUE TITLE
Add a lint rule for using `Null` as a `bool`

### ISSUE THREAD

**Issue by natebosch** [dart-lang/linter#1047]
```dart
if (someVar = null) {
   print('Not what I wanted!');
}
```

We could consider also linting if if `dynamic` is used as a `bool`.

**Comment by natebosch** [comment]
I'd also be in favor of this being an analyzer warning instead of a lint if we think we can get the false positives to zero.

**Comment by bwilkerson** [comment]
If those cover all the cases, then there isn't any reason to be more general.

**Comment by rrousselGit** [comment]
An alternative would be to have a lint rule against assignments inside condition blocks such as if/ternary/while/for
Something to prevent `if (bool = null)` but also `if ((var = new Var()).isOk)`

As this is a fairly bad practice to assign variables inside condition blocks

**Comment by pq** [comment]
@rrousselGit: I really like this proposal 👍

**Comment by natebosch** [comment]
If we can make it a Warning we wouldn't want to prevent any assignment, only look for the cases we know would fail at runtime.

If we want to limit it to a lint we could consider preventing any assignment.

**Comment by JHBitencourt** [comment]
Regarding the recent changes in the language behavior, couldn't this issue be closed now?
The analyzer now produces a [non_bool_condition](https://dart.dev/tools/diagnostic-messages#non_bool_condition) error.

**Comment by natebosch** [comment]
yeah I think this has been made obsolete by null safety.

There is likely still a risk around `dynamic`, but I think that might also be covered by `implicit-casts`.
